### PR TITLE
Documentation for DetachedMockFactory

### LINKED ISSUE #1057: Documentation for DetachedMockFactory
The `DetachedMockFactory` is only documented in the `Spring Module` section and mentioned in the Changelog.
Please add documentation for it to the normal Userguide.

### PR DESCRIPTION
**Pull request by AndreasTu**
Added a new section "Create Mock Objects outside Specifications" in the interaction_based_testing.adoc to document usage of the DetachedMockFactory. Added reference to the new section to AutoAttach.

This fixes #1057

### CODE REVIEW

File: docs/interaction_based_testing.adoc
Comment (leonard84) [inline docs/interaction_based_testing.adoc:1065]: I think we need another note, which explains that the mocks can't have any behavior unless they are attached to a spec. Or, you use a [custom IResponseGenerator](https://stackoverflow.com/a/74804706/2145769), but I'm not sure if we should include the last part as an example.

File: docs/interaction_based_testing.adoc
Comment (AndreasTu) [reply docs/interaction_based_testing.adoc]: Done

File: docs/interaction_based_testing.adoc
Comment (AndreasTu) [reply docs/interaction_based_testing.adoc:1065]: Done.

**Comment by kriegaex** [comment]
@AndreasTu, thanks for taking the initiative here. I am not the type of person who comments in a code review, kinda remote-controlling and micro-managing mypeers by telling the exactly which changes I suggest. I would rather opt to commit on top of your changes and collaborate with you that way. Would you mind giving me commit rights in your forked repo? Or would you rather have me create a forked PM, also on top of your changes? I am not a committer here, so the "Allow edits and access to secrets by maintainers" option does not have any effect on me, even though you might have activated it.

**Comment by kriegaex** [comment]
Sorry for being slow, @AndreasTu. I had noticed that you added me as a committer on your fork and started working on this PR a couple of days ago, but then was busy otherwise. Just now when trying to push my commit, I realised that I had to rebase first and that in the meantime you had also added an example for using a custom default response. My example is somewhat more verbose, also trying to create a sense of how to use it with a class under test and an injected mock. I also have an example for `@AutoAttach`. I hope that my changes are not too invasive. I think that your original PR's intent is still intact.

@leonard84, DYT it is OK to quote more code in the Spock manual? I felt like some snippets out of context might not do a proper job at explaining how to use detached mocks.

**Comment by kriegaex** [comment]
@leonard84, is there anything I can do about the false-positive code scanning errors concerning usage of the `@Shared def mockFactory` field? Some kind of exclusion comment maybe?

File: docs/extensions.adoc
Comment (leonard84) [inline docs/extensions.adoc:498]: If we add an explicit link for a pre-existing section we should use the generated one, to keep old links functioning.
`_autoattach`

File: spock-specs/src/test/groovy/org/spockframework/docs/interaction/DetachedMockFactoryDocSpec.groovy
Comment (leonard84) [inline spock-specs/src/test/groovy/org/spockframework/docs/interaction/DetachedMockFactoryDocSpec.groovy:25]: Although it works, I think we should show this order, to first attach the mock and then define behavior for it.
```suggestion
    mockUtil.attachMock(manuallyAttachedEngine, this)
    manuallyAttachedEngine.isStarted() >> true
```

File: spock-specs/src/test/groovy/org/spockframework/docs/interaction/DetachedMockFactoryDocSpec.groovy
Comment (leonard84) [inline spock-specs/src/test/groovy/org/spockframework/docs/interaction/DetachedMockFactoryDocSpec.groovy:17]: Why did you chose `static` over `@Shared`? I general I'd prefer `@Shared` over `static` whenever possible.

File: spock-specs/src/test/groovy/org/spockframework/docs/interaction/DetachedMockFactoryDocSpec.groovy
Comment (leonard84) [inline spock-specs/src/test/groovy/org/spockframework/docs/interaction/DetachedMockFactoryDocSpec.groovy:144]: I think this class should have its own `DetachedMockFactory` as it is not obvious that it is embedded in this spec, and it would be more like the real world use.

File: spock-specs/src/test/groovy/org/spockframework/docs/interaction/DetachedMockFactoryDocSpec.groovy
Comment (leonard84) [inline spock-specs/src/test/groovy/org/spockframework/docs/interaction/DetachedMockFactoryDocSpec.groovy:137]:
```suggestion
          ? ThreadLocalRandom.current().nextBoolean()
```

File: docs/interaction_based_testing.adoc
Comment (leonard84) [inline docs/interaction_based_testing.adoc:1105]: I think the primary, use-case of this would be to make the mock respond in a meaningful manner outside the context of a specification. For example if you mock something in spring and the bean is immediately used by other beans without the chance to get attached first.

And add a new line for every sentence in asciidoc, it makes diffs easier to read.

```suggestion
Usually, you would simply create your detached mock outside of Spock, but then define interactions (stubbed method results, method call verifications) inside the feature method itself.
Sometimes however, you might want to define some (user-overridable) default behaviour right in the detached mock definition itself and even make it somewhat configurable.
This is possible using a combination of Spock's mocking API and custom `IDefaultResponse` implementations.
```

File: spock-specs/src/test/groovy/org/spockframework/docs/interaction/DetachedMockFactoryDocSpec.groovy
Comment (leonard84) [inline spock-specs/src/test/groovy/org/spockframework/docs/interaction/DetachedMockFactoryDocSpec.groovy:72]: As mentioned in the text above, you should show how it behaves without being attached to the specification as well.

**Review by leonard84 — CHANGES_REQUESTED** [review]
looking good, some things to refine

File: docs/extensions.adoc
Comment (kriegaex) [reply docs/extensions.adoc:498]: I did not add this line, but I can remove it, while I am dealing with the other review findings. I am just unsure what you mean by `_autoattach`. Is that meant to be a naming suggestion or another kind of hint? For me, that is a bit too terse.

File: spock-specs/src/test/groovy/org/spockframework/docs/interaction/DetachedMockFactoryDocSpec.groovy
Comment (kriegaex) [reply spock-specs/src/test/groovy/org/spockframework/docs/interaction/DetachedMockFactoryDocSpec.groovy:25]: Absolutely, I agree. Good catch! I will also adjust the explanation in the corresponding docs section.

File: spock-specs/src/test/groovy/org/spockframework/docs/interaction/DetachedMockFactoryDocSpec.groovy
Comment (kriegaex) [reply spock-specs/src/test/groovy/org/spockframework/docs/interaction/DetachedMockFactoryDocSpec.groovy:17]: I even refactored it that way the first time, before I force-pushed the static version again. The reason is simple: The inner class `EngineMockCreator` uses the same mock factory, therefore it has to be static. Would you prefer me to make the fields shared and define a redundant mock factory inside the mock creator class?

File: spock-specs/src/test/groovy/org/spockframework/docs/interaction/DetachedMockFactoryDocSpec.groovy
Comment (kriegaex) [reply spock-specs/src/test/groovy/org/spockframework/docs/interaction/DetachedMockFactoryDocSpec.groovy:144]: Oh Ok, there is the answer to my question from the previous comment. Ok, will do.

File: spock-specs/src/test/groovy/org/spockframework/docs/interaction/DetachedMockFactoryDocSpec.groovy
Comment (kriegaex) [reply spock-specs/src/test/groovy/org/spockframework/docs/interaction/DetachedMockFactoryDocSpec.groovy:137]: Will do.

File: docs/interaction_based_testing.adoc
Comment (kriegaex) [reply docs/interaction_based_testing.adoc:1105]: Good point, I will adjust the explanation.

File: spock-specs/src/test/groovy/org/spockframework/docs/interaction/DetachedMockFactoryDocSpec.groovy
Comment (kriegaex) [reply spock-specs/src/test/groovy/org/spockframework/docs/interaction/DetachedMockFactoryDocSpec.groovy:72]: Um, well, the whole sample spec never overrides the default behaviour and uses the mocks as they are created. So either, I make the example even more verbose by overriding behaviour after attaching the mocks, or I never attach them in the first place. I think, I will do the former and also add some code comments right inside the sample spec, even though it makes the code lengthier.

Comment (kriegaex) [reply spock-specs/src/test/groovy/org/spockframework/docs/interaction/DetachedMockFactoryDocSpec.groovy]: @leonard84, can you explain why after changing the parametrised spec according to your suggestion, it is now failing? Either "ich stehe auf dem Schlauch", as we Germans like to say, and am overlooking something obvious. Or I just uncovered a Spock problem. I would like to have your insight on that. Maybe you want tu run the spec and see what it does. But for the life of me, I do not know why overriding the default behaviour does not work as I would expect it to.

File: docs/extensions.adoc
Comment (AndreasTu) [reply docs/extensions.adoc]: I guess, if you do not supply a label like `[[AutoAttach]]` [Asciidoctor](https://asciidoctor.org/) defines a default label named by the Headline which then would be `_autoattach`.
And @leonard84 meant I could have broken links to the old auto-generated label.

File: spock-specs/src/test/groovy/org/spockframework/docs/interaction/DetachedMockFactoryDocSpec.groovy
Comment (AndreasTu) [reply spock-specs/src/test/groovy/org/spockframework/docs/interaction/DetachedMockFactoryDocSpec.groovy]: This is really strange. I have added two additional tests to isolate the behavior.
`"When block interaction still active after when block without verification"`
`"When block interactions are only active during the when block, if verification present"`

The single verification line `1 * m.start()` in the `then:` block, changes the behavior of the `m.isStarted()` method for the rest of the feature execution.

Comment (AndreasTu) [reply spock-specs/src/test/groovy/org/spockframework/docs/interaction/DetachedMockFactoryDocSpec.groovy]: @kriegaex I have added a workaround to the test case.
I guess we should open up a new bug ticket to investigate the problem.

Comment (kriegaex) [reply spock-specs/src/test/groovy/org/spockframework/docs/interaction/DetachedMockFactoryDocSpec.groovy]: @AndreasTu, thanks for investigating and creating the issue. I would, however, prefer not to have the workaround pollute the demo spec and rather wait with the documentation merge after #1759 has been fixed. It would be bad to include a known problem into the docs and explain to users why the feature method looks so strange. We can still reconsider, if it proves to be difficult and time-consuming to fix.

Comment (kriegaex) [reply spock-specs/src/test/groovy/org/spockframework/docs/interaction/DetachedMockFactoryDocSpec.groovy]: Besides, I am enjoying this kind of collaborative work. @leonard84 pushed me to improve test and documentation, I uncovered unexpected Spock behaviour, @AndreasTu created a bug ticket. This is so much better than the "one person, one PR" approach and immediately proved to be valuable. 😊😊😊

**Comment by AndreasTu** [comment]
This PR is blocked until #1761 gets merged.

File: spock-specs/src/test/groovy/org/spockframework/docs/interaction/DetachedMockFactoryDocSpec.groovy
Comment (leonard84) [reply spock-specs/src/test/groovy/org/spockframework/docs/interaction/DetachedMockFactoryDocSpec.groovy]: To unblock this PR from the non trivial problem discussed in #1579, I'd suggest that we simply split this feature into two, i.e., one where the mock is attached in the given block and one where it isn't.

Comment (AndreasTu) [reply spock-specs/src/test/groovy/org/spockframework/docs/interaction/DetachedMockFactoryDocSpec.groovy]: @leonard84 Or how about my fix from before?
I reapplied it to this PR. 
This would fix the sample until #1579 gets resolved.

Comment (leonard84) [reply spock-specs/src/test/groovy/org/spockframework/docs/interaction/DetachedMockFactoryDocSpec.groovy]: IMHO having two features is the better way to showcase this anyway.

Comment (AndreasTu) [reply spock-specs/src/test/groovy/org/spockframework/docs/interaction/DetachedMockFactoryDocSpec.groovy]: I splitted the feature into two.

File: spock-specs/src/test/groovy/org/spockframework/docs/interaction/DetachedMockFactoryDocSpec.groovy
Comment (leonard84) [inline spock-specs/src/test/groovy/org/spockframework/docs/interaction/DetachedMockFactoryDocSpec.groovy:20]: We usually don't omit `def` for `@Shared` fields.
```suggestion
  @Shared 
  def mockFactory = new DetachedMockFactory()
  
  @Shared 
  def mockUtil = new MockUtil()
```

File: spock-specs/src/test/groovy/org/spockframework/docs/interaction/DetachedMockFactoryDocSpec.groovy
Comment (leonard84) [inline spock-specs/src/test/groovy/org/spockframework/docs/interaction/DetachedMockFactoryDocSpec.groovy:33]: Blocks should be separated by an empty line, like this:

```suggestion

    then:
```

File: docs/interaction_based_testing.adoc
Comment (leonard84) [inline docs/interaction_based_testing.adoc:1111]: I think we should mention here as well, that the Spring/Guice extensions will handle the attaching/detaching.

**Review by leonard84 — CHANGES_REQUESTED** [review]

File: docs/interaction_based_testing.adoc
Comment (AndreasTu) [reply docs/interaction_based_testing.adoc:1111]: I have added a sentence mentioning Spring and Guice.

**Review by leonard84 — APPROVED** [review]

**Pull request merged**
